src/cart/crt_iv.c

```diff
@@ -2564,6 +2567,82 @@ crt_ivu_rpc_issue(d_rank_t dest_rank, crt_iv_key_t *iv_key,
 	return rc;
 }
 
+static void
+handle_response_internal(void *arg)
+{
+	const struct crt_cb_info *cb_info = arg;
+	crt_rpc_t		 *rpc = cb_info->cci_rpc;
+
+	switch (rpc->cr_opc) {
+	case CRT_OPC_IV_FETCH:
+		handle_ivfetch_response(cb_info);
+		break;
+	case CRT_OPC_IV_SYNC:
+		handle_ivsync_response(cb_info);
+		break;
+	case CRT_OPC_IV_UPDATE:
+		handle_ivupdate_response(cb_info);
+		break;
+	default:
+		D_ERROR("wrong opc 0x%x\n", rpc->cr_opc);
+	}
+}
+
+static void
+handle_response_cb_internal(void *arg)
+{
+	struct crt_cb_info	*cb_info = arg;
+	crt_rpc_t		*rpc = cb_info->cci_rpc;
+	struct crt_rpc_priv	*rpc_priv;
+
+	handle_response_internal(arg);
+
+	rpc_priv = container_of(rpc, struct crt_rpc_priv, crp_pub);
+	RPC_DECREF(rpc_priv);
+	D_FREE_PTR(cb_info);
+}
+
+/* Fetch response handler */
+static void
+handle_response_cb(const struct crt_cb_info *cb_info)
+{
+	crt_rpc_t		*rpc = cb_info->cci_rpc;
+	struct crt_rpc_priv	*rpc_priv;
+	struct crt_context	*crt_ctx;
+
+	rpc_priv = container_of(rpc, struct crt_rpc_priv, crp_pub);
+	D_ASSERT(rpc_priv != NULL);
+	crt_ctx = rpc_priv->crp_pub.cr_ctx;
+
+	if (crt_rpc_cb_customized(crt_ctx, &rpc_priv->crp_pub)) {
+		int rc;
+		struct crt_cb_info *info;
+
+		D_ALLOC_PTR(info);
+		D_ASSERT(info != NULL);
+		/* NB: there are no easy way to handle allocation
+		 * failure here, since missing this callback might
+		 * cause serious problem for IV. so let's use
+		 * assert for the moment.
+		 */
+		RPC_ADDREF(rpc_priv);
+
+		info->cci_rpc = cb_info->cci_rpc;
+		info->cci_rc = cb_info->cci_rc;
+		info->cci_arg = cb_info->cci_arg;
```

Comment:
Hmm, actually here you can just reuse the original cb_info right? because all use cases in cart the cb_info is allocated.
why allocate a new info?

Reply:
and where to free the newly allocated info BTW?

Reply:
Can't just reuse cb_info passed here, since it will be auto-free-ed once the callback exits. ULT can be scheduled asynchronously so you cant guarantee it is valid once ult gets to run. Freeing is on line 2602

Reply:
> Can't just reuse cb_info passed here, since it will be auto-free-ed once the callback exits. ULT can be scheduled asynchronously so you cant guarantee it is valid once ult gets to run. Freeing is on line 2602

hmm, the auto-freeing only for the case of below L2642, no?
For the daos use case, it is safe to reuse the cb_info and make it still be freed inside handle_response_cb_internal() that will be executed by created ULT by DAOS. I did not see any problem for this usage. Or maybe I misunderstood it. @wangdi1 ?

Reply:
is  handle_response_cb_internal passed on L2635 guaranteed to be executed before this function exits? As I understood when we invoke cc_rpc_cb() the handler passed can end up executing asynchronously; if so then by the time handle_response_cb_intenral() is invoked, cb_info will already be free-ed by this function exiting

Reply:
handle_response_cb_internal() will be executed by DAOS created ULT, and is async to handle_response_cb. 
but handle_response_cb()'s exit will not cause cb_info be freed for daos' use case, no?

Reply:
it will. handle_response_cb() is the one pasesd to crt_req_send() as callback. in many instances cb_info that cb is called with is allocated on the stack of section that calls completion cb, so once it exits, you cannot use it anymore.

e.g. in crt_context.c
` 291         if (rpc_priv->crp_complete_cb != NULL) {
 292                 struct crt_cb_info      cbinfo;
 293
 294                 cbinfo.cci_rpc = &rpc_priv->crp_pub;
 295                 cbinfo.cci_arg = rpc_priv->crp_arg;
 296                 cbinfo.cci_rc = rc;
...
 310                 rpc_priv->crp_complete_cb(&cbinfo);

`

Reply:
I see what you mean. Seems the last version's detailed code made me confuse, but the updated one looks fine to me.